bls/src/main/java/tech/pegasys/teku/bls/hashToG2/Helper.java

```diff
@@ -62,65 +66,94 @@ static boolean isOnCurve(JacobianPoint p) {
     return mxChain(psi3).add(psi2).neg().add(p);
   }
 
   /**
-   * Hashes a string msg of any length into an element of the FP2 field.
+   * Produces a uniformly random byte string of arbitrary length using SHA-256.
+   *
+   * <p>As defined at https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-5.3.1
    *
-   * <p>As defined at https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-5.3
+   * @param message the message to hash
+   * @param dst the domain separation tag for the cipher suite
+   * @param lengthInBytes the number of bytes we want to obtain
+   * @return a uniformly random sequence of Bytes
+   */
+  @VisibleForTesting
+  static Bytes expandMessage(Bytes message, Bytes dst, int lengthInBytes) {
+    checkArgument(dst.size() < 256, "The DST must be 255 bytes or fewer.");
+    checkArgument(lengthInBytes > 0, "Number of bytes requested must be greater than zero.");
+
+    final int ell = 1 + (lengthInBytes - 1) / SHA256_HASH_SIZE;
+    checkArgument(ell <= 255, "Too many bytes of output were requested.");
+
+    byte[] uniformBytes = new byte[ell * SHA256_HASH_SIZE];
+
+    Bytes dstPrime = Bytes.concatenate(dst, Bytes.of((byte) dst.size()));
+    Bytes zPad = Bytes.wrap(new byte[SHA256_BLOCK_SIZE]);
+    Bytes libStr = Bytes.ofUnsignedShort(lengthInBytes);
+    Bytes b0 =
+        Hash.sha2_256(Bytes.concatenate(zPad, message, libStr, Bytes.of((byte) 0), dstPrime));
+    Bytes bb = Hash.sha2_256(Bytes.concatenate(b0, Bytes.of((byte) 1), dstPrime));
+    System.arraycopy(bb.toArrayUnsafe(), 0, uniformBytes, 0, SHA256_HASH_SIZE);
+    for (int i = 1; i < ell; i++) {
+      bb = Hash.sha2_256(Bytes.concatenate(b0.xor(bb), Bytes.of((byte) (i + 1)), dstPrime));
+      System.arraycopy(bb.toArrayUnsafe(), 0, uniformBytes, i * SHA256_HASH_SIZE, SHA256_HASH_SIZE);
+    }
+    return Bytes.wrap(uniformBytes, 0, lengthInBytes);
+  }
+
+  /**
+   * Hashes a string msg of any length into one or more elements of the FP2 field.
    *
-   * <p>This is hash_to_base() in the reference code.
+   * <p>As defined at https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-5.2
    *
    * @param message the message to hash
-   * @param ctr 0, 1, or 2 - used to efficiently create independent instances of hash_to_base
-   * @param salt key for the HMAC base hash
+   * @param count the number of field elements to return
+   * @param dst the domain separation tag for the cipher suite
    * @return an element in FP2
    */
-  static FP2Immutable hashToBase(Bytes message, byte ctr, Bytes salt) {
-
-    final Bytes h2cBytes = Bytes.wrap("H2C".getBytes(StandardCharsets.US_ASCII));
-    final Bytes ctrBytes = Bytes.of(ctr);
+  static FP2Immutable[] hashToField(Bytes message, int count, Bytes dst) {
 
-    Bytes info, t;
+    // See https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-8.8.1
+    final int l = 64;
+    // The extension degree of our field, FP2
+    final int m = 2;
```

Comment:
In the link we provided, m is set to 1. Also, at this function, it seems like we can't do use the simplified method if m = 2. https://github.com/PegaSysEng/teku/pull/1895/files#diff-67f0ab02949d5471961bcee03eb59b9bR179

Reply:
The simplified version for m=2 is shown at the end of [section 4.1](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-4.1) - that's what I'm using here. It's not possible to link to it more directly, unfortunately.